Rename `pointer` field on `Pin`

LegionMammal978 opened the pull request:
A few days ago, I was helping another user create a self-referential type using `PhantomPinned`. However, I noticed an odd behavior when I tried to access one of the type's fields via `Pin`'s `Deref` impl:

```rust
use std::{marker::PhantomPinned, ptr};

struct Pinned {
    data: i32,
    pointer: *const i32,
    _pin: PhantomPinned,
}

fn main() {
    let mut b = Box::pin(Pinned {
        data: 42,
        pointer: ptr::null(),
        _pin: PhantomPinned,
    });
    {
        let pinned = unsafe { b.as_mut().get_unchecked_mut() };
        pinned.pointer = &pinned.data;
    }
    println!("{}", unsafe { *b.pointer });
}
```

```rust
error[E0658]: use of unstable library feature 'unsafe_pin_internals'
  --> <source>:19:30
   |
19 |     println!("{}", unsafe { *b.pointer });
   |                              ^^^^^^^^^

error[E0277]: `Pinned` doesn't implement `std::fmt::Display`
  --> <source>:19:20
   |
19 |     println!("{}", unsafe { *b.pointer });
   |                    ^^^^^^^^^^^^^^^^^^^^^ `Pinned` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Pinned`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Since the user named their field `pointer`, it conflicts with the `pointer` field on `Pin`, which is public but unstable since Rust 1.60.0 with #93176. On versions from 1.33.0 to 1.59.0, where the field on `Pin` is private, this program compiles and prints `42` as expected.

To avoid this confusing behavior, this PR renames `pointer` to `__pointer`, so that it's less likely to conflict with a `pointer` field on the underlying type, as accessed through the `Deref` impl. This is technically a breaking change for anyone who names their field `__pointer` on the inner type; if this is undesirable, it could be renamed to something more longwinded. It's also a nightly breaking change for any external users of `unsafe_pin_internals`.

LegionMammal978 commented:
@rustbot label +T-libs-api -T-libs

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7403746316/job/20144117953) [(plain)](https://github.com/rust-lang/rust/commit/0750d0d85babf59f367be911f0e1983d3fbc7b3e/checks/20144117953/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=LegionMammal978
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_7898f2bf-1159-4fcd-8672-bacbf3f02bbf
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=6a91c88223847b9f800503025ddfb530142d5a69
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_7898f2bf-1159-4fcd-8672-bacbf3f02bbf
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_7898f2bf-1159-4fcd-8672-bacbf3f02bbf
GITHUB_TRIGGERING_ACTOR=LegionMammal978
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/119562/merge
GITHUB_WORKFLOW_SHA=6a91c88223847b9f800503025ddfb530142d5a69
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Jan  4 00:04:55 UTC 2024
  network time: Thu, 04 Jan 2024 00:04:55 GMT
  network time: Thu, 04 Jan 2024 00:04:55 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/feature-gates/feature-gate-unsafe_pin_internals.rs stdout ----
diff of stderr:

11 LL | #![forbid(internal_features, unsafe_code)]
13 
- error: aborting due to 1 previous error
- error: aborting due to 1 previous error
+ error[E0560]: struct `Pin<&mut T>` has no field named `pointer`
+    |
+    |
+ LL |     Pin { pointer }
+    |           ^^^^^^^ `Pin<_>` does not have this field
+    = note: all struct fields are already assigned
15 
+ error: aborting due to 2 previous errors
+ 
---
   = note: using it is strongly discouraged
note: the lint level is defined here
  --> /checkout/tests/ui/feature-gates/feature-gate-unsafe_pin_internals.rs:2:11
   |
LL | #![forbid(internal_features, unsafe_code)]


error[E0560]: struct `Pin<&mut T>` has no field named `pointer`
   |
   |
LL |     Pin { pointer }
   |           ^^^^^^^ `Pin<_>` does not have this field
   = note: all struct fields are already assigned

error: aborting due to 2 previous errors

```

</details>

Nilstrieb commented on library/core/src/pin.rs:
```diff
@@ -414,11 +414,11 @@ pub struct Pin<P> {
     #[unstable(feature = "unsafe_pin_internals", issue = "none")]
     #[doc(hidden)]
-    pub pointer: P,
+    pub __pointer: P,
```
can you leave a comment mentioning the deref impact of the field?

LegionMammal978 replied:
I've added a comment.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7411621209/job/20166560244) [(plain)](https://github.com/rust-lang/rust/commit/48fbe38f69300813e2adf4c8ab63f76d9b426bf9/checks/20166560244/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=LegionMammal978
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_21273299-5752-468a-a1da-7d7112c14ff6
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c356dbe9ca8d3b0cca1b594d38d90de0031b6936
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_21273299-5752-468a-a1da-7d7112c14ff6
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_21273299-5752-468a-a1da-7d7112c14ff6
GITHUB_TRIGGERING_ACTOR=LegionMammal978
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/119562/merge
GITHUB_WORKFLOW_SHA=c356dbe9ca8d3b0cca1b594d38d90de0031b6936
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Jan  4 15:42:18 UTC 2024
  network time: Thu, 04 Jan 2024 15:42:18 GMT
  network time: Thu, 04 Jan 2024 15:42:18 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
..................

failures:

---- [mir-opt] tests/mir-opt/inline/inline_coroutine.rs stdout ----
36 -         _4 = g() -> [return: bb1, unwind continue];
37 +         _4 = {coroutine@$DIR/inline_coroutine.rs:19:5: 19:8 (#0)};
38 +         _3 = &mut _4;
- +         _2 = Pin::<&mut {coroutine@$DIR/inline_coroutine.rs:19:5: 19:8}> { pointer: move _3 };
+ +         _2 = Pin::<&mut {coroutine@$DIR/inline_coroutine.rs:19:5: 19:8}> { __pointer: move _3 };
40 +         StorageDead(_3);
41 +         StorageLive(_5);
42 +         _5 = const false;
thread '[mir-opt] tests/mir-opt/inline/inline_coroutine.rs' panicked at src/tools/compiletest/src/runtest.rs:4104:21:
Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/inline/inline_coroutine.main.Inline.panic-unwind.diff
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #119722) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 985402dbe20e27a897a2899a4e83630fb13b01c3 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 985402dbe20e27a897a2899a4e83630fb13b01c3 -->
<!-- homu: {"type":"Approved","sha":"985402dbe20e27a897a2899a4e83630fb13b01c3","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
I think this failed in https://github.com/rust-lang/rust/pull/119755#issuecomment-1882277975

LegionMammal978 commented:
I've updated the expression in `src/etc/natvis/libcore.natvis` to account for the change in field name. I believe it should work, but I haven't tested it. Anyway, I've been thinking it might be better to hold off on merging this PR until #119615 is resolved one way or another, since that PR would then also have to revert this PR, and all the tests would have been touched for nothing.

bors commented:
:umbrella: The latest upstream changes (presumably #119672) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

LegionMammal978 committed:
Rename `pointer` field on `Pin`

The internal, unstable field of `Pin` can conflict with fields from the
inner type accessed via the `Deref` impl. Rename it from `pointer` to
`__pointer`, to make it less likely to conflict with anything else.

dtolnay commented:
@bors r=Amanieu,dtolnay

bors commented:
:pushpin: Commit bc3fb5245a248070b76e535be4f342e2be12a9b5 has been approved by `Amanieu,dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu,dtolnay bc3fb5245a248070b76e535be4f342e2be12a9b5 -->
<!-- homu: {"type":"Approved","sha":"bc3fb5245a248070b76e535be4f342e2be12a9b5","approver":"Amanieu,dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

LegionMammal978 commented:
@dtolnay Has #119615 been decided against? As I said earlier, I'd like to avoid needless reverts if that ends up as the solution.

dtolnay commented:
I am personally on board with #119615, but it has been waiting 2 weeks for feedback from petrochenkov, and I don't think it makes sense to block this fix on that. Inserting a revert into that PR later is not a big deal.
